### PR TITLE
ENH: DataFrame.style sparsified MultiIndex

### LINKED ISSUE #11655: Conditional HTML styling hides MultiIndex structure
I noticed that the styling doesn't quite work with `MultiIndex` data frames. For example if we have a dataframe with two cross validated scores and take their `mean`and `sem` using `groupby` to get a `MultiIndex` dataframe

    import itertools
    import pandas as pd
    import numpy as np

    jobs = itertools.product(['a', 'b', 'c', 'd'], np.arange(1e-4, 1e-3, .0003), range(10))
    rows = []
    for v, v2, itr in jobs:
        rows.append({'param_1': v, 'param_2': v2, 'iter': itr,
                     'score_1': np.random.randint(0, 100, size=(1,))[0],
                     'score_2': np.random.rand(1, )[0]})
    df_multi = pd.DataFrame(rows)
    agg = df_multi.groupby(by=['param_1', 'param_2'])[['score_1', 'score_2']].agg(['mean', 'sem'])
    agg.style.highlight_max()

screen shot 2015-11-20 at 12 04 13

where as without the styling the index names are preserved and the `score_1` and `score_2` column headers span 2 columns which improves legibility.

screen shot 2015-11-20 at 12 02 56

### PR DESCRIPTION
**Pull request by TomAugspurger**
- [x] closes #11655
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

[Notebook comparing `DataFrame._html_repr_` to `DataFrame.style`](https://gist.github.com/609c398f814b4a505bf4f406670e457e)

I think we're identical for non-truncated DataFrames. That' has not been implemented in `Styler` yet.

Along the way I noticed two other things that ended up needing fixing.
1. DataFrame.columns.names were not displayed
2. CSS classes weren't being assigned correctly to row labels.

The fixes ended up being pretty intertwined, so I've put them in a single PR. Unfortunately, the commits are a bit jumbled as well :/

### CODE REVIEW

File: doc/source/whatsnew/v0.19.0.txt [hunk 1]
```diff
@@ -314,6 +314,8 @@ Other enhancements
 - ``Series.append`` now supports the ``ignore_index`` option (:issue:`13677`)
 - ``.to_stata()`` and ``StataWriter`` can now write variable labels to Stata dta files using a dictionary to make column names to labels (:issue:`13535`, :issue:`13536`)
 - ``.to_stata()`` and ``StataWriter`` will automatically convert ``datetime64[ns]`` columns to Stata format ``%tc``, rather than raising a ``ValueError`` (:issue:`12259`)
+- ``DataFrame.style`` will now render sparsified MultiIndexes (:issue:`11655`)
+- ``DataFrame.style`` will now show column names (:issue:``)
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: use the pr number

File: doc/source/whatsnew/v0.19.0.txt [hunk 2]
```diff
@@ -761,3 +763,4 @@ Bug Fixes
 - Bug where ``pd.read_gbq()`` could throw ``ImportError: No module named discovery`` as a result of a naming conflict with another python package called apiclient  (:issue:`13454`)
 - Bug in ``Index.union`` returns an incorrect result with a named empty index (:issue:`13432`)
 - Bugs in ``Index.difference`` and ``DataFrame.join`` raise in Python3 when using mixed-integer indexes (:issue:`13432`, :issue:`12814`)
+- Bug in the CSS classes assigned to ``DataFrame.style`` for index names. Previously they were assigned ``"col_heading level<n> col<c>"`` where ``n`` was the number of levels + 1. Now they are assigned ``"index_name level<n>"``, where ``n`` is the correct level for that MultiIndex.
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: maybe put a ref to the style docs (e.g. not sure how much of an example is needed here)

**Comment by TomAugspurger** [comment]
> maybe put a ref to the style docs (e.g. not sure how much of an example is needed here)

@jreback this was undocumented... Adding this section to the notebook. Will push soon.

# CSS Classes

Certain CSS classes are attached to cells.

- Index and Column names include `index_name` and `level<k>` where `k` is its level in a MultiIndex
- Index label cells include
  + `row_heading`
  + `row<n>` where `n` is the numeric position of the row
  + `level<k>` where `k` is the level in a MultiIndex
- Column label cells include
  + `col_heading`
  + `col<n>` where `n` is the numeric position of the column
  + `level<k>` where `k` is the level in a MultiIndex
- Blank cells include `blank`
- Data cells include `data`

File: doc/source/whatsnew/v0.19.0.txt [hunk 3]
```diff
@@ -314,6 +314,8 @@ Other enhancements
 - ``Series.append`` now supports the ``ignore_index`` option (:issue:`13677`)
 - ``.to_stata()`` and ``StataWriter`` can now write variable labels to Stata dta files using a dictionary to make column names to labels (:issue:`13535`, :issue:`13536`)
 - ``.to_stata()`` and ``StataWriter`` will automatically convert ``datetime64[ns]`` columns to Stata format ``%tc``, rather than raising a ``ValueError`` (:issue:`12259`)
+- ``DataFrame.style`` will now render sparsified MultiIndexes (:issue:`11655`)
+- ``DataFrame.style`` will now show column names (:issue:`13775`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.19.0.txt]: 'column names' -> 'column level names' ? ('column names' sounds like the columns labels itself)

Comment (TomAugspurger) [inline doc/source/whatsnew/v0.19.0.txt]: 👍  ``DataFrame.style`` will now show column level names (e.g. ``DataFrame.columns.names``) (:issue:`13775`)

File: doc/source/html-styling.ipynb [hunk 4]
```diff
@@ -792,6 +792,27 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
+    "# CSS Classes\n",
+    "\n",
+    "Certain CSS classes are attached to cells.\n",
+    "\n",
+    "- Index and Column names include `index_name` and `level<k>` where `k` is its level in a MultiIndex\n",
```
Comment (jreback) [inline doc/source/html-styling.ipynb]: would add this documenation to the doc-string as well

File: pandas/formats/style.py [hunk 5]
```diff
@@ -893,6 +933,40 @@ def _highlight_extrema(data, color='yellow', max_=True):
...
 
+def _is_visible(idx_row, idx_col, lengths):
+    """
+    Index -> {(idx_row, idx_col): bool})
+    """
+    return (idx_col, idx_row) in lengths
+
+
+def _get_level_lengths(index):
+    '''
```
Comment (jreback) [inline pandas/formats/style.py]: use triple-quotes

File: pandas/tests/formats/test_style.py [hunk 6]
```diff
@@ -8,11 +8,6 @@
 from pandas.util.testing import TestCase
 import pandas.util.testing as tm
 
-# this is a mess. Getting failures on a python 2.7 build with
-# whenever we try to import jinja, whether it's installed or not.
-# so we're explicitly skipping that one *before* we try to import
-# jinja. We still need to export the imports as globals,
-# since importing Styler tries to import jinja2.
 job_name = os.environ.get('JOB_NAME', None)
```
Comment (jreback) [inline pandas/tests/formats/test_style.py]: does this still happen?

File: pandas/tests/formats/test_style.py [hunk 7]
```diff
@@ -581,6 +594,136 @@ def f(x):
         with tm.assertRaises(ValueError):
             df.style._apply(f, axis=None)
 
+    def test_get_level_lengths(self):
+        index = pd.MultiIndex.from_product([['a', 'b'], [0, 1, 2]])
+        expected = {(0, 0): 3, (0, 3): 3, (1, 0): 1, (1, 1): 1, (1, 2): 1,
+                    (1, 3): 1, (1, 4): 1, (1, 5): 1}
+        result = _get_level_lengths(index)
+        self.assertDictEqual(result, expected)
```
Comment (jreback) [inline pandas/tests/formats/test_style.py]: use ``tm.assert_dict_equal``

File: pandas/tests/formats/test_style.py (hunk 6 above)
Comment (TomAugspurger) [inline pandas/tests/formats/test_style.py]: Oops, meant to just delete the first sentence. I don't know if it's still a problem, will test it out in a followup.

**Comment by jreback** [comment]
lgtm. merge away.

**Pull request closed without merging**